### ISSUE TITLE
Double Escaping &nbsp; and &reg; When Using OptionOutputAsXml

### ISSUE THREAD

**Issue by mbrierst** [zzzprojects/html-agility-pack#83]
When using OptionOutputAsXml and loading html, \&nbsp; ends up double escaped as \&amp;nbsp.  \&reg; is similar.  At least some other entities (like \&lt;, \&quot;) do not have this behavior.  Workarounds involving replacing text before processing and before saving are possible.
Here is a demo of the problem in C#:
```C#
var htmlDocument = new HtmlDocument { OptionOutputAsXml = true /* false for normal behavior */};
htmlDocument.LoadHtml("&nbsp;&lt;&gt;&reg;&quot;&amp;");
MessageBox.Show(htmlDocument.DocumentNode.InnerHtml);
```

**Comment by JonathanMagnan** [comment]
Hello @mbrierst ,

Thank you for reporting, I will investigate this issue today.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
The issue is caused by the method `HtmlEncode` in HAP,

I'm not sure if there is a problem or not with this scenario. 

In fact, there is some problem with the `HtmlEncode` method since it will not support a `HtmlDecode` by looking how it has been coded.

The following string `"&nbsp;&lt;&gt;&reg;&quot;&amp;"` normally get encoded in XML as `&amp;nbsp;&amp;lt;&amp;gt;&amp;reg;&amp;quot;&amp;amp;`
See: https://www.freeformatter.com/xml-escape.html

Something it's sure is `&nbsp;` should be formatted as `&amp;nbsp;` since the `&` need to be escaped in XML.

Do you understand why this is hapenning with my answer and why this is not an issue?

As said, there is some issue in the method depending if we should thread already escaped character such as `&lt;` as already being escaped or escape it again to support `HtmlDecode`.

Let me know if I'm wrong with my explanation.

Best Regards,

Jonathan

**Comment by mbrierst** [comment]
I'm not sure I understand what you're saying.  I definitely think the current behavior is inconsistent, and is making it more difficult to write correct code.

If I load an XHtml document `&lt;` I expect the result to have InnerHtml `&lt;`, and if I then write out that same document, I get that same result, and everything makes sense.

If I load an XHtml document `&reg;` I expect the result to have InnerHtml `&reg;`, but it instead shows `&amp;reg`.  When I write out the document, I get `&amp;reg`, which does not match the document I read in.

So it seems to me that the behavior for `&reg;` (and `&nbsp;`) is wrong, and the behavior for `&lt;` is correct.  This belief is reinforced by the fact that when I use `OptionOutputAsXml = false` I get consistent behavior just like `&lt;`.

**Comment by JonathanMagnan** [comment]
Let me explain this another way,

If the library support `Escape`, the client should be able to retrieve the initial message with `Unescape`

So, 

if the initial input is `'&'`
- The library must return the following escaped string `&amp;`
- The client can now unescape the string back to `'&'`

if the initial input is `'<'`
- The library must return the following escaped string `&lt;`
- The client can now unescape the string back to `'<'`

if the initial input is `'<&lt;'`
- The library must return the following escaped string `&lt;&amp;lt`
- The client can now unescape the string back to `'<&lt;'`

If we follow your example and the initial input is `'<&lt;'`
- You want the library to return `'&lt;&lt;'`
- And the client will unescape it to '<<'

which is **NOT** equal to the initial input

Here is the problem with your logic, it make impossible for the client to unescape and return the original string.

Let me know if that's more clear.

Best Regards,

Jonathan

**Comment by mbrierst** [comment]
I understand what you are saying, it all makes sense and your logic is correct.  I don't want the behavior you describe to change.  But that's not quite what I'm talking about.  I don't believe any escaping or unescaping should be happening in this example at all, or if it is happening it should be invisible to the user.

Let's start with a file that is just `&nbsp;&lt;`.  When displayed on a web page, most browsers would show it as a non breaking space followed by a "<".  I'll load this into HAP, and then write the file back out (I'm simplifying all this by just loading and showing strings, but the effect is the same if you use files):
 ```C#
var d = new HtmlDocument { OptionOutputAsXml = false };
d.LoadHtml("&nbsp;&lt;");
MessageBox.Show(d.DocumentNode.InnerHtml);
```
This all works just like I would expect.  The string at the end matches the string I loaded.  No problem.

Now, just change one word, changing OptionOutputAsXml to true:
 ```C#
var d = new HtmlDocument { OptionOutputAsXml = true };
d.LoadHtml("&nbsp;&lt;");
MessageBox.Show(d.DocumentNode.InnerHtml);
```
Suddenly the output string is NOT the same as the input string was, it is now `&amp;nbsp;&lt;`.  All I did was change the OptionOutputAsXml option.  The &lt; is still escaped the same way as in the initial string.  But \&nbsp; is escaped differently from how it was in the initial string.  So there is no consistent way to even fix the string.  If I unescape it, the result string will be `&nbsp;<`, which is still not correct.

**Comment by JonathanMagnan** [comment]
There is for sure one problem as we both recognize. Currently they is no way to unescape correctly what has been escaped by the library since some already escaped character such as `&lt;` are not escaped as `&amp;lt;`

---
The problem is `&nbsp;` cannot stay `&nbsp;` otherwise it can break some HTML.

See character to escape: https://www.freeformatter.com/xml-escape.html

- `'` is replaced with `&apos;`
- `"` is replaced with `&quot;`
- `&` is replaced with `&amp;`
- `<` is replaced with `&lt;`
- `>` is replaced with `&gt;`

In short `&lt;` is valid in a XML file but not `&nbsp;` because it can break XML, that's why this is escaped and not `&lt;`. One is valid in XML, the other no.

**Comment by mbrierst** [comment]
I see what you're saying, but in some cases \&nbsp; IS valid XML, if you use the correct DOCTYPE.  For example, it is valid XHTML, which is a strict XML format, and which is what I was trying to output.  For example, run this file:
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1 plus MathML 2.0//EN" "http://www.w3.org/Math/DTD/mathml2/xhtml-math11-f.dtd">
<html>
<head><title>Title</title></head>
<body><div>&nbsp;</div></body>
</html>
```
through either of these validators (one is for HTML and one is for XML), it will pass:
https://validator.w3.org/#validate_by_input+with_options
https://www.w3schools.com/xml/xml_validator.asp
If you run through either validator WITHOUT the DOCTYPE it will fail.  If HAP has no real concept of what  DOCTYPEs mean, I think assuming XHTML during OutputAsXml is a good choice, as that's a likely XML output from an HTML parser.

**Comment by JonathanMagnan** [comment]
Thank you for the additional info,

I will check this,

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @mbrierst ,

I understand your point and you may be proven me wrong.

One problem stay that a lot of people already use the library with the current logic and adding this kind of fix have very impact on existing code.

At one point, I need to find some way to be able to work less on the 1.x version and more on the 2.x version.

To make everyone happy, I have added a new option `BackwardCompatibility` that when turning it to `false`. some behavior like the output is not guaranteed to be the same since some fix could be applied.

By example,

```csharp
var htmlDocument = new HtmlAgilityPack.HtmlDocument { BackwardCompatibility = false, OptionOutputAsXml = true /* false for normal behavior */};
htmlDocument.LoadHtml("&nbsp;&lt;&gt;&reg;&quot;&amp;");
var s = htmlDocument.DocumentNode.InnerHtml;
```

Will now return your expected string.

This change should be released before next Monday.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @mbrierst ,

The v1.6.1 has been released.

Please reopen this issue if more change is required.

Best Regards,

Jonathan

**Comment by mbrierst** [comment]
Yep, looks like that fixed it for me!  Thanks!
